File: test/test_safe_replace.js
#!/usr/bin/env node
/**
 * Unit tests for safe description-only replacement logic
 * Verifies that localization ONLY replaces description:"..." fields
 * and does NOT break === comparisons, ternary expressions, case statements, etc.
 */

const assert = require('assert');

// ========== Import safe_replace from localize.js ==========
// We inline the function since localize.js modifies files directly
function escapeRegex(str) {
    return str.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
}

function safeReplace(src, keyword, translation) {
    let count = 0;
    const escaped = escapeRegex(keyword);
    const descDoubleRegex = new RegExp(`(description:")(${escaped})(")`, 'g');
    src = src.replace(descDoubleRegex, (match, prefix, content, suffix) => {
        count++;
        return `${prefix}${translation}${suffix}`;
    });
    const descTmplRegex = new RegExp(`(description:\`)(${escaped})(\`)`, 'g');
    src = src.replace(descTmplRegex, (match, prefix, content, suffix) => {
        count++;
        return `${prefix}${translation}${suffix}`;
    });
    return { src, count };
}

// ========== Test Cases ==========
let passed = 0;
let failed = 0;

function test(name, fn) {
    try {
        fn();
        passed++;
        console.log(`  \x1b[32mPASS\x1b[0m ${name}`);
    } catch (err) {
        failed++;
        console.log(`  \x1b[31mFAIL\x1b[0m ${name}`);
        console.log(`       ${err.message}`);
    }
}

console.log('\n\x1b[95m=== Safe Replace Unit Tests ===\x1b[0m\n');

// ===== SHOULD replace (description fields) =====

test('Replaces description:"keyword" with translation', () => {
    const src = 'description:"Cancel"';
    const { src: result, count } = safeReplace(src, 'Cancel', '取消');
    assert.strictEqual(result, 'description:"取消"');
    assert.strictEqual(count, 1);
});

test('Replaces description:`keyword` (template literal)', () => {
    const src = 'description:`Cancel`';
    const { src: result, count } = safeReplace(src, 'Cancel', '取消');
    assert.strictEqual(result, 'description:`取消`');
    assert.strictEqual(count, 1);
});

test('Replaces multiple occurrences in description fields', () => {
    const src = 'description:"Cancel" other stuff description:"Cancel"';
    const { src: result, count } = safeReplace(src, 'Cancel', '取消');
    assert.strictEqual(result, 'description:"取消" other stuff description:"取消"');
    assert.strictEqual(count, 2);
});

test('Replaces long multi-word descriptions', () => {
    const src = 'description:"Enter to submit · Esc to cancel"';
    const { src: result, count } = safeReplace(src, 'Enter to submit · Esc to cancel', 'Enter 提交 · Esc 取消');
    assert.strictEqual(result, 'description:"Enter 提交 · Esc 取消"');
    assert.strictEqual(count, 1);
});

test('Replaces description with special regex chars', () => {
    const src = 'description:"Fast mode (${Ok} only)"';
    const { src: result, count } = safeReplace(src, 'Fast mode (${Ok} only)', '快速模式（仅限 ${Ok}）');
    assert.strictEqual(result, 'description:"快速模式（仅限 ${Ok}）"');
    assert.strictEqual(count, 1);
});

// ===== SHOULD NOT replace (code logic) =====

test('Does NOT replace === comparison', () => {
    const src = 'if(M6==="Language")return"en"';
    const { src: result, count } = safeReplace(src, 'Language', '语言');
    assert.strictEqual(result, 'if(M6==="Language")return"en"');
    assert.strictEqual(count, 0);
});

test('Does NOT replace ternary expression', () => {
    const src = 'C6==="enable"?"Enabled":"Disabled"';
    const { src: result, count } = safeReplace(src, 'Enabled', '已启用');
    assert.strictEqual(result, 'C6==="enable"?"Enabled":"Disabled"');
    assert.strictEqual(count, 0);
});

test('Does NOT replace standalone quoted string', () => {
    const src = 'label:"Cancel"';
    const { src: result, count } = safeReplace(src, 'Cancel', '取消');
    assert.strictEqual(result, 'label:"Cancel"');
    assert.strictEqual(count, 0);
});

test('Does NOT replace variable name', () => {
    const src = 'const Cancel = document.getElementById("cancel")';
    const { src: result, count } = safeReplace(src, 'Cancel', '取消');
    assert.strictEqual(result, 'const Cancel = document.getElementById("cancel")');
    assert.strictEqual(count, 0);
});

test('Does NOT replace string in array', () => {
    const src = '["Cancel","Back","Submit"]';
    const { src: result, count } = safeReplace(src, 'Cancel', '取消');
    assert.strictEqual(result, '["Cancel","Back","Submit"]');
    assert.strictEqual(count, 0);
});

test('Does NOT replace object key value', () => {
    const src = 'name:"Cancel",value:"cancel"';
    const { src: result, count } = safeReplace(src, 'Cancel', '取消');
    assert.strictEqual(result, 'name:"Cancel",value:"cancel"');
    assert.strictEqual(count, 0);
});

test('Does NOT replace in switch case', () => {
    const src = 'case"Cancel":handleCancel();break';
    const { src: result, count } = safeReplace(src, 'Cancel', '取消');
    assert.strictEqual(result, 'case"Cancel":handleCancel();break');
    assert.strictEqual(count, 0);
});

test('Does NOT replace in function call argument', () => {
    const src = 't("Cancel")';
    const { src: result, count } = safeReplace(src, 'Cancel', '取消');
    assert.strictEqual(result, 't("Cancel")');
    assert.strictEqual(count, 0);
});

test('Does NOT replace in property assignment', () => {
    const src = 'title="Cancel";';
    const { src: result, count } = safeReplace(src, 'Cancel', '取消');
    assert.strictEqual(result, 'title="Cancel";');
    assert.strictEqual(count, 0);
});

// ===== EDGE CASES =====

test('Empty input returns unchanged', () => {
    const { src: result, count } = safeReplace('', 'Cancel', '取消');
    assert.strictEqual(result, '');
    assert.strictEqual(count, 0);
});

test('No match returns unchanged', () => {
    const src = 'description:"Something Else"';
    const { src: result, count } = safeReplace(src, 'Cancel', '取消');
    assert.strictEqual(result, 'description:"Something Else"');
    assert.strictEqual(count, 0);
});

test('Similar prefix does not false match', () => {
    const src = 'description:"Cancel" description:"Cancel item"';
    const { src: result, count } = safeReplace(src, 'Cancel', '取消');
    assert.strictEqual(count, 1);
    assert(result.includes('description:"取消"'));
    assert(result.includes('description:"Cancel item"'));
});

// ===== REAL-WORLD CLUSTER TEST =====

test('Real minified code cluster: only description: gets replaced', () => {
    const src = `M6==="Language"?M6:void 0;description:"Language"===g6?g6:void 0;name:"Language"`;
    const { src: result, count } = safeReplace(src, 'Language', '语言');
    // Only description:"Language" should be replaced
    assert.strictEqual(count, 1);
    assert(result.includes('M6==="Language"'));       // NOT replaced
    assert(result.includes('description:"语言"'));      // replaced
    assert(result.includes('name:"Language"'));       // NOT replaced
});

// ========== Summary ==========
console.log(`\n\x1b[95m=== Results ===\x1b[0m`);
console.log(`  Passed: ${passed}`);
console.log(`  Failed: ${failed}`);

if (failed > 0) {
    console.log('\n\x1b[31mTESTS FAILED!\x1b[0m');
    process.exit(1);
} else {
    console.log('\n\x1b[32mAll tests passed!\x1b[0m\n');
}

File: test/test_safe_replace.py
#!/usr/bin/env python3
"""
Unit tests for Python safe_replace function
Verifies description-only replacement strategy
"""

import re
import sys

PASSED = 0
FAILED = 0

def escape_regex(s):
    return re.escape(s)

def safe_replace(content, keyword, translation):
    """Same logic as localize.py"""
    count = 0
    escaped = re.escape(keyword)
    pattern_double = r'(description:")(' + escaped + r')(")'
    content, n = re.subn(pattern_double, r'\g<1>' + translation + r'\g<3>', content)
    count += n
    pattern_tmpl = r'(description:`)(' + escaped + r')(`)'
    content, n = re.subn(pattern_tmpl, r'\g<1>' + translation + r'\g<3>', content)
    count += n
    return content, count

def test(name, fn):
    global PASSED, FAILED
    try:
        fn()
        PASSED += 1
        print(f"  \033[92mPASS\033[0m {name}")
    except AssertionError as e:
        FAILED += 1
        print(f"  \033[91mFAIL\033[0m {name}")
        print(f"       {e}")

print("\n\033[95m=== Python safe_replace Unit Tests ===\033[0m\n")

# Should replace
test('Replaces description:"keyword"', lambda: (
    safe_replace('description:"Cancel"', 'Cancel', '取消')
    .__eq__(('description:"取消"', 1)) or _assert()
) if False else None)

def _assert(): raise AssertionError("fail")

# Manual test style to avoid lambda issues
r, c = safe_replace('description:"Cancel"', 'Cancel', '取消')
if r == 'description:"取消"' and c == 1:
    PASSED += 1; print(f"  \033[92mPASS\033[0m Replace description:\"keyword\"")
else:
    FAILED += 1; print(f"  \033[91mFAIL\033[0m Replace description:\"keyword\"")

r, c = safe_replace('description:`Cancel`', 'Cancel', '取消')
if r == 'description:`取消`' and c == 1:
    PASSED += 1; print(f"  \033[92mPASS\033[0m Replace description:`keyword`")
else:
    FAILED += 1; print(f"  \033[91mFAIL\033[0m Replace description:`keyword`")

# Should NOT replace
r, c = safe_replace('if(M6==="Language")return"en"', 'Language', '语言')
if '==="Language")' in r and c == 0:
    PASSED += 1; print(f"  \033[92mPASS\033[0m Skip === comparison")
else:
    FAILED += 1; print(f"  \033[91mFAIL\033[0m Skip === comparison")

r, c = safe_replace('C6==="enable"?"Enabled":"Disabled"', 'Enabled', '已启用')
if '"Enabled"' in r and c == 0:
    PASSED += 1; print(f"  \033[92mPASS\033[0m Skip ternary expression")
else:
    FAILED += 1; print(f"  \033[91mFAIL\033[0m Skip ternary expression")

r, c = safe_replace('label:"Cancel"', 'Cancel', '取消')
if 'label:"Cancel"' in r and c == 0:
    PASSED += 1; print(f"  \033[92mPASS\033[0m Skip label:\"...\"")
else:
    FAILED += 1; print(f"  \033[91mFAIL\033[0m Skip label:\"...\"")

r, c = safe_replace('name:"Language"', 'Language', '语言')
if 'name:"Language"' in r and c == 0:
    PASSED += 1; print(f"  \033[92mPASS\033[0m Skip name:\"...\"")
else:
    FAILED += 1; print(f"  \033[91mFAIL\033[0m Skip name:\"...\"")

# Real-world cluster
r, c = safe_replace('M6==="Language"?M6:void 0;description:"Language"', 'Language', '语言')
if c == 1 and '==="Language")' not in r and 'description:"语言"' in r:
    PASSED += 1; print(f"  \033[92mPASS\033[0m Real code cluster test")
else:
    FAILED += 1; print(f"  \033[91mFAIL\033[0m Real code cluster test")

# Summary
print(f"\n\033[95m=== Results ===\033[0m")
print(f"  Passed: {PASSED}")
print(f"  Failed: {FAILED}")

if FAILED > 0:
    print("\n\033[91mTESTS FAILED!\033[0m")
    sys.exit(1)
else:
    print("\n\033[92mAll tests passed!\033[0m\n")
